benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,225 @@
+"""Classes for persisting benchmark results."""
+
+from __future__ import annotations
+
+import json
+import os
+from datetime import datetime
+from enum import Enum
+from pathlib import Path
+from typing import Protocol
+
+import boto3
+import boto3.session
+from attr import define, field
+from attrs.validators import instance_of
+from boto3.session import Session
+from git import Repo
+from typing_extensions import override
+
+from benchmarks import Benchmark, Result
+
+VARNAME_BENCHMARKING_PERSISTENCE_PATH = "BAYBE_BENCHMARKING_PERSISTENCE_PATH"
+
+
+class PathStrategy(Enum):
+    """Specifies the way a file path is constructed."""
+
+    HIERARCHICAL = "HIERARCHICAL"
+    """Hierarchical path construction using folders."""
+
+    FLAT = "FLAT"
+    """Flat path construction using a file name only."""
+
+
+@define
+class PathConstructor:
+    """A class to construct the file path of a result object.
+
+    The class encapsulates the construction of a file path depending
+    on where the object should be stored. Since different storage backends have
+    different requirements for the path, and some like the used S3 bucket
+    do not support folders, the path uses its variables to construct the path
+    in the order of the variables with a separator depending on the strategy.
+
+    For compatibility reasons, the path is sanitized to contain only the following
+    allowed characters:
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+    """
+
+    benchmark_name: str = field(validator=instance_of(str))
+    """The name of the benchmark for which the path should be constructed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    latest_baybe_tag: str = field(validator=instance_of(str))
+    """The latest BayBE version tag."""
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    commit_hash: str = field(validator=instance_of(str))
```

Comment:
RE branch and hash I just realized: this should not be the branch/hash that is **currently** checked out but rather at the time of the result creation, right? So it makes sense to me that the `commit` is populated based on the result, but why is this not the case for the branch? I mean, I see that the branch info is not stored in the result, but why does it then appear here? Is it relevant or not?

Reply:
You are correct, what applies to the commit hash should also apply to the branch. I think it is relevant information to filter and should be conceptually treated the same as the commit hash. Therefore, I will move it to the result metadata.

Reply:
Ok 👍🏼  Can you then please also rephrase the docstrings accordingly? I.e. `... checked out during benchmark execution`. Also, while the "the latest BayBE version tag" works in the MetaData class, it doesn't work here since the creation time of the objects are different, i.e. you can't copy-paste the docstring. That is, there could theoretically be a more recent version tag already. So you need to explicitly say `The latest BayBE version tag at benchmark execution time.`

Reply:
Okay, done.